File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations
def solution(k, dungeons):
    init_k = k
    answer = -1
    for i in list(permutations(dungeons, len(dungeons))):
        tmp = i
        cnt = 0
        k = init_k
        for t in tmp:
            if k >= t[0]:
                k -= t[1]
                cnt += 1
            else:
                break
        
        if cnt == len(dungeons):
            answer = cnt
            break
        answer = max(answer, cnt)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {

    static int maxCount = 0;

    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        dfs(k, dungeons, visited, 0);
        return maxCount;
    }

    public void dfs(int k, int[][] dungeons, boolean[] visited, int count) {
        maxCount = Math.max(maxCount, count);

        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(k - dungeons[i][1], dungeons, visited, count + 1);
                visited[i] = false;
            }
        }
    }
}